core/templates/dev/head/forms/formBuilder.js

```diff
@@ -747,9 +726,383 @@ oppia.config(['$provide', function($provide) {
+                   * need to finalizeCreation again).
+                   */
+                  if (!that.isReady()) {
+                    // Actually insert the widget, if we have not already.
+                    editor.widgets.finalizeCreation(container);
+                  }
+
+                  /**
+                   * Need to manually $compile so the directive renders.
+                   * Note that.element.$ is the native DOM object
+                   * represented by that.element. See:
+                   * http://docs.ckeditor.com/#!/api/CKEDITOR.dom.element
+                   */
+                  $compile($(that.element.$).contents())($rootScope);
+                  // $timeout ensures we do not take the undo snapshot until
+                  // after angular finishes its changes to the component tags
+                  $timeout(function() {
+                    // For inline widgets, place the caret after the
+                    // widget so the user can continue typing immediately.
+                    if (isInline) {
+                      var range = editor.createRange();
+                      var widgetContainer = that.element.getParent();
+                      range.moveToPosition(
+                        widgetContainer, CKEDITOR.POSITION_AFTER_END);
+                      editor.getSelection().selectRanges([range]);
+                      // Another timeout needed so the undo snapshot is
+                      // not taken until the caret is in the right place.
+                      $timeout(function() {
+                        editor.fire('unlockSnapshot');
+                        editor.fire('saveSnapshot');
+                      });
+                    } else {
+                      editor.fire('unlockSnapshot');
+                      editor.fire('saveSnapshot');
+                    }
+                  });
+                },
+                function() {},
+                function() {});
+            },
+            /**
+             * This is how the widget will be represented in the outputs source,
+             * so it is called when we call editor.getData().
+             */
+            downcast: function(element) {
+              // Clear the angular rendering content, which we don't
+              // want in the output.
+              element.children[0].setHtml('');
+              // Return just the rich text component, without its wrapper.
+              return element.children[0];
+            },
+            /**
+             * This is how a widget is recognized by CKEditor, for example
+             * when we first load data in. Returns a boolean,
+             * true iff "element" is an instance of this widget.
+             */
+            upcast: function(element) {
+              return (element.name !== 'p' &&
+                      element.children.length > 0 &&
+                      element.children[0].name === tagName);
+            },
+            data: function() {
```

Comment:
When is this called? Suggest adding docstring.

(It feels slightly weird to me that the method name has no verb and that it returns nothing -- but again, perhaps that's just how it is.)